### PR TITLE
Convert VariableSet to RealType only

### PR DESCRIPTION
**Pull request by camelto2**
## Proposed changes

While working to enable complex orbital rotations, @rcclay, @jptowns and I realized there were some inconsistencies in how we deal with optimizable parameters and their types. 

The optimizer is hard coded deal with only real parameters since it takes the real part of the parameters stored in VariableSet.  Generally, to minimize a **complex analytic** function you only need to consider d/dz, but general functions that aren't complex analytic need to be formulated with d/dz and d/dz*.  For those types of functions, it is completely equivalent to work with just d/d(re(z)) and d/d(im(z)) though. The local energy and cost function that we minimize are not complex analytic, so we need to work with both d/d(re(z)) and d/d(im(z)). So, to make the optimizer properly deal with complex parameters, can keep the optimizer as is if we just treat the real and imaginary part of complex numbers as two independent real parameters. So the optimizer wouldn't need to be touched.

The interface between the optimizer and optimizable wave function components is VariableSet, and it currently stored everything as ValueType. Since the optimizer only ever needs to deal with reals (treating the real/imaguinary components as 2 reals), the VariableSet should only store reals. Currently, no existing wave function component actually supports optimizing a complex parameter, so changing VariableSet to be RealType only is not a breaking change. 

If any future wave function component wants to support optimizing a complex parameter, it needs to simply register the complex parameter as two reals into VariableSet, and then the optimizer will work as it is currently written. We are going to submit a future PR enabling this functionality for RotatedSPOs. 

This PR simply makes the change in VariableSet to remove ValueType, and change the various places in the code where it was used. The existing unit test for VariableSet is modified to remove the checks on ValueType, so these changes are already being tested with existing code. If there are any additional tests that should be added, let me know. 

## What type(s) of changes does this code introduce?
_Delete the items that do not apply_

- Refactoring (no functional changes, no api changes)
- Testing changes (e.g. new unit/integration/performance tests)

### Does this introduce a breaking change?
- No. Note that complex MSD CI coefficients optimization is broken as it is.

## What systems has this change been tested on?
macOS

## Checklist

_Update the following with a yes where the items apply. If you're unsure about any of them, don't hesitate to ask.  This is
simply a reminder of what we are going to look for before merging your code._

- Yes. This PR is up to date with current the current state of 'develop'
- Yes. Code added or changed in the PR has been clang-formatted
- No. This PR adds tests to cover any new code, or to catch a bug that is being fixed
- No. Documentation has been added (if appropriate)

### CODE REVIEW

**Comment by ye-luo** [comment]
This change also get the logic of mixing real/complex parameters in TWF straight. Real parameters like in Jastrow factors remain real and complex parameters are casted as two real parameters.

File: src/QMCDrivers/WFOpt/QMCCostFunctionBase.cpp [hunk 1]
```diff
@@ -251,7 +251,7 @@ void QMCCostFunctionBase::reportParametersH5()
   if (!myComm->rank())
   {
     int ci_size = 0;
-    std::vector<opt_variables_type::value_type> CIcoeff;
+    std::vector<opt_variables_type::real_type> CIcoeff;
```
Comment (ye-luo) [inline src/QMCDrivers/WFOpt/QMCCostFunctionBase.cpp:254]: This is part is not so right in math but that is the current state of the code. Optimized CI coefficients should remain in complex. So optimizing complex CI coefficients needs fix following the complexification of RotatedSPO. Could you open an issue tracking this.

File: src/QMCWaveFunctions/Fermion/SlaterDetBuilder.cpp [hunk 2]
```diff
@@ -561,10 +561,10 @@ std::unique_ptr<MultiSlaterDetTableMethod> SlaterDetBuilder::createMSDFast(
     Optimizable = CI_Optimizable = true;
     if (csf_data_ptr)
       for (int i = 1; i < csf_data_ptr->coeffs.size(); i++)
-        myVars.insert(CItags[i], csf_data_ptr->coeffs[i], true, optimize::LINEAR_P);
+        myVars.insert(CItags[i], std::real(csf_data_ptr->coeffs[i]), true, optimize::LINEAR_P);
     else
       for (int i = 1; i < C.size(); i++)
-        myVars.insert(CItags[i], C[i], true, optimize::LINEAR_P);
+        myVars.insert(CItags[i], std::real(C[i]), true, optimize::LINEAR_P);
```
Comment (ye-luo) [inline src/QMCWaveFunctions/Fermion/SlaterDetBuilder.cpp:567]: Will need a fix for the complex CI coefficients optimization later.

**Review by ye-luo — APPROVED** [review]

**Comment by ye-luo** [comment]
Test this please

File: src/QMCDrivers/WFOpt/QMCCostFunctionBase.cpp (hunk 1 above)
Comment (camelto2) [reply src/QMCDrivers/WFOpt/QMCCostFunctionBase.cpp:254]: done

Comment (ye-luo) [reply src/QMCDrivers/WFOpt/QMCCostFunctionBase.cpp:254]: https://github.com/QMCPACK/qmcpack/issues/4820

**Pull request merged**